<?xml version="1.0"?>
<!--
  Copyright 2015 Haoyao Chen HITSZ Shenzhen/ASL, ETH Zurich, Switzerland

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
<!-- size_link v0 0.025 v1 0.0025-->
<!--	v0:<box size="0.01 0.01 ${length_stick}" />   v1:<box size="0.001 0.001 ${length_stick}" />  -->
<robot xmlns:xacro="http://ros.org/wiki/xacro">

  <!--inter_link for suspended stick/cable-->
  <xacro:macro name="suspended_sphere_link"
	  params="namespace name_link mass_link size_link *origin">
    <link name="${namespace}/${name_link}" gravity="true">
			<inertial>
			  <mass value="${mass_link}"/>
			  <inertia 
			  	ixx="0.00100000"
			  	ixy="0.000000"
					ixz="0.000000"
					iyy="0.0010000"
					iyz="0.000000"
					izz="0.0010000">
				</inertia>
				<xacro:insert_block name="origin" />
			</inertial>
			<collision name="sphere_01_collision">
			  <geometry>
					<sphere
						radius="0.0025">
					</sphere>
				</geometry>
				<xacro:insert_block name="origin" />
			</collision>
			<visual name="sphere_01_vis">
		    <geometry>
		      <sphere
						radius="${size_link}">
					</sphere>
			  </geometry>
       	<material name="red" >
					<color rgba="0 0.2 0.0 0.0" />
				</material>
				<xacro:insert_block name="origin" />
			</visual>
			<velocity_decay />
		</link>
  </xacro:macro>

	<!--body_link for suspended stick/cable-->
  <xacro:macro name="suspended_body_link"
		params="namespace link_index mass_stick length_stick *origin">
		<link name="${namespace}/suspended_stick_body_link${link_index}" gravity="false">
			<inertial>
			  <mass value="0.000100000"/>
			  <inertia 
			  	ixx="0.00100000"
			  	ixy="0.000000"
					ixz="0.000000"
					iyy="0.0010000"
					iyz="0.000000"
					izz="0.0010000">
				</inertia>
				<xacro:insert_block name="origin" />
			</inertial>

      <!--xacro:box_inertial
				x="0.01"
				y="0.01"
				z="${length_stick}"
				mass="${mass_stick}">
				<xacro:insert_block name="origin" />
      </xacro:box_inertial-->
     <collision>
			  <geometry>
			    <box size="0.001 0.001 ${length_stick}" />
			  </geometry>
				<xacro:insert_block name="origin" />
			</collision>
			<visual>
	      <geometry>
			    <box size="0.001 0.001 ${length_stick}" />
	      </geometry>
        <material name="red" >
					<color rgba="0 0.9 0.0 0.0" />
				</material>
				<xacro:insert_block name="origin" />
	    </visual>
    </link>
    <gazebo reference="${namespace}/suspended_stick_body_link${link_index}">
			<gravity>false</gravity>
		</gazebo>
	</xacro:macro>	 

  <!-- Macro to add a suspended stick which connects to the payload. -->
  <xacro:macro name="suspended_stick_plugin_macro"
    params="namespace parent_link mass_stick length_stick *origin">
		<!--suspended stick base-head joint -->
   	<joint name="${namespace}/suspended_cable_base_head_joint" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${parent_link}" />
      	<child link="${namespace}/suspended_head_link"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <!-- suspended stick head link -->
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_head_link"	>
			<origin xyz="0 0 0.0" rpy="0 0 0" />
	  </xacro:suspended_sphere_link>
    <!--suspended stick head-body joint -->
   	<joint name="${namespace}/suspended_stick_head_body_joint" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_head_link" />
      	<child link="${namespace}/suspended_stick_body_link"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<!-- suspended stick body link -->
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index=""
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
    <!--suspended stick body-tail joint -->
   	<joint name="${namespace}/suspended_stick_body_tail_joint" type="revolute">
				<origin xyz="0.0 0.0 ${-length_stick}" rpy="0.0 0.0 0.0" />
      	<parent link="${namespace}/suspended_stick_body_link" />
      	<child link="${namespace}/suspended_tail_link"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.5" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_tail_link"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
  </xacro:macro>

  <!-- Macro to add a suspended cable simulated by several connected rigid sticks. -->
  <xacro:macro name="suspended_cable_plugin_macro"
    params="namespace parent_link mass_stick length_stick mass_tail_link size_tail_link *origin">
		<!--suspended stick base-head joint -->
   	<joint name="${namespace}/suspended_stick_base_head_joint" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${parent_link}" />
      	<child link="${namespace}/suspended_head_link"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
    <!-- suspended stick head link -->
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_head_link"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
   	<joint name="${namespace}/suspended_cable_head_body_joint" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_head_link" />
      	<child link="${namespace}/suspended_stick_body_link1"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0.0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="1"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
		<joint name="${namespace}/suspended_cable_inter_joint11" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link1" />
      	<child link="${namespace}/suspended_sphere_link1"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link1"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint12" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link1" />
      	<child link="${namespace}/suspended_stick_body_link2"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="2"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint22" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link2" />
      	<child link="${namespace}/suspended_sphere_link2"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link2"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint23" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link2" />
      	<child link="${namespace}/suspended_stick_body_link3"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="3"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint33" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link3" />
      	<child link="${namespace}/suspended_sphere_link3"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link3"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint34" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link3" />
      	<child link="${namespace}/suspended_stick_body_link4"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="4"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint44" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link4" />
      	<child link="${namespace}/suspended_sphere_link4"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link4"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint45" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link4" />
      	<child link="${namespace}/suspended_stick_body_link5"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="5"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint55" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link5" />
      	<child link="${namespace}/suspended_sphere_link5"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link5"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint56" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link5" />
      	<child link="${namespace}/suspended_stick_body_link6"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="6"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint66" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link6" />
      	<child link="${namespace}/suspended_sphere_link6"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link6"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint67" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link6" />
      	<child link="${namespace}/suspended_stick_body_link7"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="7"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint77" type="revolute">
				<origin xyz="0 0 -${length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link7" />
      	<child link="${namespace}/suspended_sphere_link7"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link7"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint78" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link7" />
      	<child link="${namespace}/suspended_stick_body_link8"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="8"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint88" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link8" />
      	<child link="${namespace}/suspended_sphere_link8"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link8"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint89" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link8" />
      	<child link="${namespace}/suspended_stick_body_link9"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="9"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint99" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link9" />
      	<child link="${namespace}/suspended_sphere_link9"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="0.0005"
			size_link="0.0025"
		  name_link="suspended_sphere_link9"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
 		<joint name="${namespace}/suspended_cable_inter_joint910" type="revolute">
				<origin xyz="0 0 0.0" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_sphere_link9" />
      	<child link="${namespace}/suspended_stick_body_link10"  />
        <axis xyz="0 1 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
	  <xacro:suspended_body_link
			namespace="${namespace}" 
			link_index="10"
			mass_stick="${mass_stick}"
			length_stick="${length_stick}" >
  		<origin xyz="0.0 0.0 ${-length_stick * 0.5}" rpy="0.0 0.0 0.0" />
    </xacro:suspended_body_link>
  	<joint name="${namespace}/suspended_stick_inter_joint1010" type="revolute">
				<origin xyz="0 0 ${-length_stick}" rpy="0 0 0" />
      	<parent link="${namespace}/suspended_stick_body_link10" />
      	<child link="${namespace}/suspended_tail_link"  />
        <axis xyz="1 0 0" />
				<dynamics damping="0.01" friction="0.0" />
        <limit upper="3.1" lower="-3.1" effort="0" velocity="0" />
   	</joint>
		<xacro:suspended_sphere_link
			namespace="${namespace}"
			mass_link="${mass_tail_link}"
			size_link="${size_tail_link}"
		  name_link="suspended_tail_link"	>
  		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	  </xacro:suspended_sphere_link>
        
  </xacro:macro>	
</robot>
